File: translator/fasta.py
def parse_fasta(text: str) -> list[tuple[str, str]]:
    """Parse FASTA text, or treat plain text as a single unnamed sequence."""
    records: list[tuple[str, str]] = []
    header: str | None = None
    chunks: list[str] = []

    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line:
            continue
        if line.startswith(">"):
            if header is not None or chunks:
                records.append((header or "sequence", "".join(chunks)))
            header = line[1:].strip() or "sequence"
            chunks = []
        else:
            chunks.append(line)

    if header is not None or chunks:
        records.append((header or "sequence", "".join(chunks)))

    return records

File: web/src/lib/fasta.ts
export type FastaRecord = {
  header: string;
  sequence: string;
};

export function parseFasta(text: string): FastaRecord[] {
  const records: FastaRecord[] = [];
  let header: string | null = null;
  let chunks: string[] = [];

  for (const rawLine of text.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line) {
      continue;
    }
    if (line.startsWith(">")) {
      if (header !== null || chunks.length > 0) {
        records.push({
          header: header ?? "sequence",
          sequence: chunks.join(""),
        });
      }
      header = line.slice(1).trim() || "sequence";
      chunks = [];
    } else {
      chunks.push(line);
    }
  }

  if (header !== null || chunks.length > 0) {
    records.push({
      header: header ?? "sequence",
      sequence: chunks.join(""),
    });
  }

  return records;
}

